Design/Rule0054DatabaseOperationInTryFunctions.cs

```diff
@@ -0,0 +1,44 @@
+using Microsoft.Dynamics.Nav.CodeAnalysis;
+using Microsoft.Dynamics.Nav.CodeAnalysis.Diagnostics;
+using Microsoft.Dynamics.Nav.CodeAnalysis.InternalSyntax;
+using Microsoft.Dynamics.Nav.CodeAnalysis.Syntax;
+using System.Collections.Immutable;
+
+// This rule is not finished 100% yet.
+// It does false warning if the a function is called like one of the database functions, but it is not a database function.
+// it should also analyze the references recursively to see if the current statement is inside a try function or not.
+
+namespace BusinessCentral.LinterCop.Design
+{
+    [DiagnosticAnalyzer]
+    public class Rule0054DatabaseOperationInTryFunctions : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(DiagnosticDescriptors.Rule0054DatabaseOperationInTryFunctions);
+
+        private static readonly List<string> databaseInvocations = new List<string>
+        {
+            "insert", "delete", "modify", "modifyall", "rename", "addlink", "deletelink", "deletelinks", "commit"
```

Comment:
add locktable

Reply:
If we need to take `Rec.LockTable()` into account, then we also should include [Rec.ReadIsolation](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-read-isolation)?

I do have a question about this do

> Because changes made to the database by a try function are not rolled back, you should not include database write transactions within a try function

https://learn.microsoft.com/en-us/previous-versions/dynamicsnav-2018-developer/Handling-Errors-by-Using-Try-Functions#DbWriteTransactions

My first though is that a `Rec.LockTable()` doesn't created changes that could case issues when not rolled back. Do you have more insights or an example where this could cause an issue?

Reply:
I believe that below code should raise this warning

```
Rec.ReadIsolation := IsolationLevel::UpdLock;
Rec..FindFirst()
```

That is, IsolationLevel::UpdLock only causes a transaction if a read operation is called afterwards.

This can be even more complex, since the LockTable() or ReadIsolation could be called before the TryFunction is invoked, and a Find* operation is done inside the TryFunction. But I think that we could ignore that scenario, and focus on the 99.9% of cases 😅

Reply:
But maybe that kind of transaction is no problem for TryFunctions? 🤔
I need to set up an environment with the DisableWriteInsideTryFunctions set, to repro this...